File: python/src/cosmoner/apps.py
"""Apps service namespace — finding apps and rolling image apps onto a new image."""

from __future__ import annotations

import asyncio
import re
import time
from collections.abc import Callable
from typing import Any

from ._config import ClientConfig, resolve_project_id
from ._transport import AsyncTransport, Transport

#: Phases after which a deployment will not change again.
FINISHED_DEPLOYMENT_PHASES = ("ACTIVE", "ERROR", "CANCELED", "SUPERSEDED")

DEFAULT_POLL_INTERVAL = 3.0
DEFAULT_WAIT_TIMEOUT = 600.0

# Docker's tag grammar and the digest form, matching what the API accepts, so a
# malformed value fails before it spends a request against the deploy budget.
_TAG_PATTERN = re.compile(r"^[A-Za-z0-9_][A-Za-z0-9_.-]{0,127}$")
_DIGEST_PATTERN = re.compile(r"^sha256:[a-f0-9]{64}$")


def _deploy_payload(app_id: str, tag: str | None, digest: str | None) -> dict[str, Any]:
    """Validates deploy arguments and shapes them into the API request body."""
    _require_app_id(app_id)
    if tag and digest:
        raise ValueError("Pass either tag or digest, not both")
    if tag is not None and not _TAG_PATTERN.fullmatch(tag):
        raise ValueError(f'Invalid image tag "{tag}"')
    if digest is not None and not _DIGEST_PATTERN.fullmatch(digest):
        raise ValueError("digest must be sha256:<64 hex characters>")

    payload: dict[str, Any] = {}
    if tag is not None:
        payload["tag"] = tag
    if digest is not None:
        payload["digest"] = digest

    return payload


def _require_app_id(app_id: str) -> None:
    """Rejects an empty app id before it becomes a malformed route."""
    if not app_id:
        raise ValueError("app_id is required")


def _require_deployment_ids(app_id: str, deployment_id: str) -> None:
    """Rejects empty app or deployment ids before they become a malformed route."""
    _require_app_id(app_id)
    if not deployment_id:
        raise ValueError("deployment_id is required")


def _validate_wait(interval: float, timeout: float) -> None:
    """Rejects polling settings that would spin or never start."""
    if interval <= 0:
        raise ValueError("interval must be greater than 0")
    if timeout <= 0:
        raise ValueError("timeout must be greater than 0")


def _timeout_error(deployment_id: str, phase: str, timeout: float) -> TimeoutError:
    """Builds the error raised when a deployment outlives the wait."""
    return TimeoutError(
        f"Deployment {deployment_id} was still {phase} after {round(timeout)}s"
    )


class AppsService:
    """Synchronous app lookup and image deploy operations for a project."""

    def __init__(self, transport: Transport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        return f"/v1/projects/{resolve_project_id(self._config, project_id)}/apps"

    def list(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Lists every app in the project, newest first."""
        result: dict[str, Any] = self._transport.request(
            "GET", self._base_path(project_id)
        )
        return result

    def deploy(
        self,
        app_id: str,
        *,
        tag: str | None = None,
        digest: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Starts deploying an image app and returns the deployment without waiting.

        Pass ``tag`` or ``digest`` to roll onto that image from the repository the
        app already pulls from, or neither to re-resolve the image the app names
        now. Only image apps pulling from a Cosmoner registry can be deployed this
        way. Pair with :meth:`wait_for_deployment` to learn whether it succeeded.
        """
        payload = _deploy_payload(app_id, tag, digest)

        result: dict[str, Any] = self._transport.request(
            "POST", f"{self._base_path(project_id)}/{app_id}/deployments", json=payload
        )
        return result

    def get_deployment(
        self, app_id: str, deployment_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Fetches one deployment's current phase."""
        _require_deployment_ids(app_id, deployment_id)

        result: dict[str, Any] = self._transport.request(
            "GET",
            f"{self._base_path(project_id)}/{app_id}/deployments/{deployment_id}",
        )
        return result

    def wait_for_deployment(
        self,
        app_id: str,
        deployment_id: str,
        *,
        interval: float = DEFAULT_POLL_INTERVAL,
        timeout: float = DEFAULT_WAIT_TIMEOUT,
        on_poll: Callable[[dict[str, Any]], None] | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Polls a deployment until it finishes, and returns it in its final phase.

        Returns for every finished phase, failures included — check ``phase`` for
        ``ACTIVE``. Raises only when a request fails or ``timeout`` seconds pass
        first (``TimeoutError``), in which case the deployment keeps going
        server-side. ``on_poll`` receives every poll result, including the last.
        """
        _validate_wait(interval, timeout)
        deadline = time.monotonic() + timeout

        while True:
            deployment: dict[str, Any] = self.get_deployment(
                app_id, deployment_id, project_id=project_id
            )["data"]
            if on_poll is not None:
                on_poll(deployment)
            if deployment["phase"] in FINISHED_DEPLOYMENT_PHASES:
                return deployment

            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise _timeout_error(deployment_id, deployment["phase"], timeout)
            time.sleep(min(interval, remaining))


class AsyncAppsService:
    """Asynchronous counterpart to :class:`AppsService`."""

    def __init__(self, transport: AsyncTransport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        return f"/v1/projects/{resolve_project_id(self._config, project_id)}/apps"

    async def list(self, *, project_id: str | None = None) -> dict[str, Any]:
        """Lists every app in the project, newest first."""
        result: dict[str, Any] = await self._transport.request(
            "GET", self._base_path(project_id)
        )
        return result

    async def deploy(
        self,
        app_id: str,
        *,
        tag: str | None = None,
        digest: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Starts deploying an image app and returns the deployment without waiting.

        Pass ``tag`` or ``digest`` to roll onto that image from the repository the
        app already pulls from, or neither to re-resolve the image the app names
        now. Only image apps pulling from a Cosmoner registry can be deployed this
        way. Pair with :meth:`wait_for_deployment` to learn whether it succeeded.
        """
        payload = _deploy_payload(app_id, tag, digest)

        result: dict[str, Any] = await self._transport.request(
            "POST", f"{self._base_path(project_id)}/{app_id}/deployments", json=payload
        )
        return result

    async def get_deployment(
        self, app_id: str, deployment_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Fetches one deployment's current phase."""
        _require_deployment_ids(app_id, deployment_id)

        result: dict[str, Any] = await self._transport.request(
            "GET",
            f"{self._base_path(project_id)}/{app_id}/deployments/{deployment_id}",
        )
        return result

    async def wait_for_deployment(
        self,
        app_id: str,
        deployment_id: str,
        *,
        interval: float = DEFAULT_POLL_INTERVAL,
        timeout: float = DEFAULT_WAIT_TIMEOUT,
        on_poll: Callable[[dict[str, Any]], None] | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Polls a deployment until it finishes, and returns it in its final phase.

        Returns for every finished phase, failures included — check ``phase`` for
        ``ACTIVE``. Raises only when a request fails or ``timeout`` seconds pass
        first (``TimeoutError``), in which case the deployment keeps going
        server-side. ``on_poll`` receives every poll result, including the last.
        """
        _validate_wait(interval, timeout)
        deadline = time.monotonic() + timeout

        while True:
            response = await self.get_deployment(
                app_id, deployment_id, project_id=project_id
            )
            deployment: dict[str, Any] = response["data"]
            if on_poll is not None:
                on_poll(deployment)
            if deployment["phase"] in FINISHED_DEPLOYMENT_PHASES:
                return deployment

            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise _timeout_error(deployment_id, deployment["phase"], timeout)
            await asyncio.sleep(min(interval, remaining))

File: javascript/src/services/apps.ts
/* eslint-disable require-await -- Every method here validates its arguments before
   reaching the transport. Keeping them `async` makes a bad argument reject the
   returned promise rather than throw synchronously, so one `.catch()` covers both
   client-side and server-side failures. */

/** Apps service namespace — finding apps and rolling image apps onto a new image. */

import { resolveProjectId, type ResolvedConfig } from "../config";
import type { Transport } from "../transport";

/** Lifecycle state of an app as a whole. */
export type AppStatus = "DEPLOYING" | "RUNNING" | "STOPPED" | "ERROR" | "TERMINATED";

/** What a registry push does to an image app. */
export type ImageDeployPolicy = "TAG" | "NEWEST" | "MANUAL";

/**
 * An app as the API lists it.
 *
 * Only the fields a deploy script is likely to read are typed; the API returns
 * more, and they are present on the object at runtime.
 */
export interface App {
  id: string;
  name: string;
  subdomain: string;
  status: AppStatus;
  url: string | null;
  /** Set for apps built from a repository, null for image apps. */
  gitRepo: string | null;
  /** The image an image app runs, null for apps built from a repository. */
  containerImage: string | null;
  imageDeployPolicy: ImageDeployPolicy;
  createdAt: string;
  updatedAt: string;
}

/** Phase of one deployment. */
export type DeploymentPhase =
  | "PENDING"
  | "BUILDING"
  | "DEPLOYING"
  | "ACTIVE"
  | "ERROR"
  | "CANCELED"
  | "SUPERSEDED";

/** Phases after which a deployment will not change again. */
export const FINISHED_DEPLOYMENT_PHASES: readonly DeploymentPhase[] = [
  "ACTIVE",
  "ERROR",
  "CANCELED",
  "SUPERSEDED",
];

/** One deployment of an app. */
export interface AppDeployment {
  id: string;
  phase: DeploymentPhase;
  /** Why it happened, e.g. "registry push" or "api deploy". */
  cause: string | null;
  /** The image reference that was deployed. */
  imageRef: string | null;
  imageDigest: string | null;
  error: string | null;
  startedAt: string;
  finishedAt: string | null;
}

/** Options accepted by every method, for working across projects. */
export interface ProjectScopedParams {
  /** Overrides the client-level default project for this call. */
  projectId?: string;
}

/**
 * Arguments accepted by `client.apps.deploy()`.
 *
 * Pass `tag` or `digest` to roll onto that image from the repository the app
 * already pulls from, or neither to re-resolve the image the app names now.
 */
export interface DeployAppParams extends ProjectScopedParams {
  tag?: string;
  /** `sha256:` followed by 64 hex characters. */
  digest?: string;
}

/** Options accepted by `client.apps.waitForDeployment()`. */
export interface WaitForDeploymentParams extends ProjectScopedParams {
  /** Milliseconds between polls. Defaults to 3s. */
  interval?: number;
  /** Milliseconds to wait in total before giving up. Defaults to 10 minutes. */
  timeout?: number;
  /** Called with every poll result, including the last. */
  onPoll?: (deployment: AppDeployment) => void;
}

interface Envelope<T> {
  success: true;
  data: T;
}

export type ListAppsResponse = Envelope<App[]>;
export type DeployAppResponse = Envelope<AppDeployment>;
export type GetDeploymentResponse = Envelope<AppDeployment>;

const DEFAULT_POLL_INTERVAL_MS = 3_000;
const DEFAULT_WAIT_TIMEOUT_MS = 10 * 60_000;

// Docker's tag grammar and the digest form, matching what the API accepts, so a
// malformed value fails before it spends a request against the deploy budget.
const TAG_PATTERN = /^[A-Za-z0-9_][A-Za-z0-9_.-]{0,127}$/;
const DIGEST_PATTERN = /^sha256:[a-f0-9]{64}$/;

/** App lookup and image deploy operations for a project. */
export class AppsService {
  constructor(
    private readonly transport: Transport,
    private readonly config: ResolvedConfig
  ) {}

  /** Builds the collection route for the resolved project. */
  private basePath(projectId?: string): string {
    return `/v1/projects/${resolveProjectId(this.config, projectId)}/apps`;
  }

  /** Lists every app in the project, newest first. */
  async list(params: ProjectScopedParams = {}): Promise<ListAppsResponse> {
    return this.transport.request<ListAppsResponse>("GET", this.basePath(params.projectId));
  }

  /**
   * Starts deploying an image app and returns the deployment without waiting.
   *
   * Only image apps pulling from a Cosmoner registry can be deployed this way;
   * the API rejects apps built from a repository. Pair with
   * `waitForDeployment()` to learn whether the rollout succeeded.
   */
  async deploy(appId: string, params: DeployAppParams = {}): Promise<DeployAppResponse> {
    if (!appId) throw new Error("appId is required");
    if (params.tag && params.digest) throw new Error("Pass either tag or digest, not both");
    if (params.tag !== undefined && !TAG_PATTERN.test(params.tag)) {
      throw new Error(`Invalid image tag "${params.tag}"`);
    }
    if (params.digest !== undefined && !DIGEST_PATTERN.test(params.digest)) {
      throw new Error("digest must be sha256:<64 hex characters>");
    }

    return this.transport.request<DeployAppResponse>(
      "POST",
      `${this.basePath(params.projectId)}/${appId}/deployments`,
      { body: { tag: params.tag, digest: params.digest } }
    );
  }

  /** Fetches one deployment's current phase. */
  async getDeployment(
    appId: string,
    deploymentId: string,
    params: ProjectScopedParams = {}
  ): Promise<GetDeploymentResponse> {
    if (!appId) throw new Error("appId is required");
    if (!deploymentId) throw new Error("deploymentId is required");

    return this.transport.request<GetDeploymentResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${appId}/deployments/${deploymentId}`
    );
  }

  /**
   * Polls a deployment until it finishes, and returns it in its final phase.
   *
   * Resolves for every finished phase, failures included — check `phase` for
   * `ACTIVE`. Rejects only when the request fails or `timeout` passes first,
   * in which case the deployment keeps going server-side.
   */
  async waitForDeployment(
    appId: string,
    deploymentId: string,
    params: WaitForDeploymentParams = {}
  ): Promise<AppDeployment> {
    const interval = params.interval ?? DEFAULT_POLL_INTERVAL_MS;
    const timeout = params.timeout ?? DEFAULT_WAIT_TIMEOUT_MS;
    if (interval <= 0) throw new Error("interval must be greater than 0");
    if (timeout <= 0) throw new Error("timeout must be greater than 0");

    const deadline = Date.now() + timeout;

    for (;;) {
      const { data } = await this.getDeployment(appId, deploymentId, params);
      params.onPoll?.(data);
      if (FINISHED_DEPLOYMENT_PHASES.includes(data.phase)) return data;

      const remaining = deadline - Date.now();
      if (remaining <= 0) {
        throw new Error(
          `Deployment ${deploymentId} was still ${data.phase} after ${Math.round(timeout / 1000)}s`
        );
      }
      await new Promise((resolve) => setTimeout(resolve, Math.min(interval, remaining)));
    }
  }
}
